hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java

```diff
@@ -0,0 +1,406 @@
+import com.hazelcast.test.annotation.SlowTest;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+import org.junit.runners.Parameterized.UseParametersRunnerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static com.hazelcast.client.impl.clientside.ClientTestUtil.getHazelcastClientInstanceImpl;
+import static com.hazelcast.jet.core.JobStatus.RUNNING;
+import static com.hazelcast.jet.core.JobStatus.SUSPENDED;
+import static com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl.EVENT_SYNC_FREQUENCY_PROP;
+import static com.hazelcast.spi.impl.eventservice.impl.EventServiceTest.getEventService;
+import static com.hazelcast.test.OverridePropertyRule.set;
+import static java.util.Arrays.asList;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toList;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParametrizedRunner.class)
+@UseParametersRunnerFactory(HazelcastSerialParametersRunnerFactory.class)
+@Category(SlowTest.class)
+public class JobStatusListenerTest extends SimpleTestInClusterSupport {
+    private static final Function<String, String> SIMPLIFY = log -> log.replaceAll("(?<=\\().*: ", "");
+    private static final String MAP_NAME = "runCount";
+
+    @Parameters(name = "{0}")
+    public static Iterable<Object[]> parameters() {
+        return asList(
+                mode("onMaster", () -> instances()[0]),
+                mode("onNonMaster", () -> instances()[1]),
+                mode("onClient", () -> client()));
+    }
+
+    static Object[] mode(String name, Supplier<HazelcastInstance> supplier) {
+        return new Object[] {name, supplier};
+    }
+
+    /**
+     * Used to generate unique job ids to be used inside jobs, e.g. sinks,
+     * since it is not easy to access the context.
+     */
+    private static int nextJobId;
+
+    @Parameter(0)
+    public String mode;
+
+    @Parameter(1)
+    public Supplier<HazelcastInstance> instance;
+
+    private String jobIdString;
+    private UUID registrationId;
+
+    /** Disable asynchronous transmission of events to preserve their order. */
+    @ClassRule
+    public static OverridePropertyRule eventSyncFrequency = set(EVENT_SYNC_FREQUENCY_PROP, "1");
```

Comment:
`EventService` sends a synchronous event at every `EVENT_SYNC_FREQUENCY_PROP`th event, for which it retransmits the event until a response is received or the maximum retry count is reached. Other events are sent asynchronously, i.e. without waiting for responses, which sometimes leads to out-of-order events on other members. Setting `EVENT_SYNC_FREQUENCY_PROP` to 1 ensures that all transmissions are synchronous. However, I don't know why we don't use the TCP's builtin packet ordering logic, so that even asynchronous transmission won't break the order.

Reply:
> which sometimes leads to out-of-order events

can this happen for normal usage of `JobStatusListener` by the users as well? If so, do we recommend setting `EVENT_SYNC_FREQUENCY_PROP` to 1 (IMO risky idea)? Out-of-order job state change events would be wrong.

Reply:
> we don't use the TCP's builtin packet ordering logic

If I am correct, it is not the TCP or lack of TCP but `EventServiceImpl.accept` that handles incoming event packets in parallel. It uses `StripedExecutor` (so they should be ordered for the same `orderKey`) but with `orderKey = packet.getPartitionId()` which seems wrong to me.

Reply:
hmm, the name is misleading, you used `jobId` as `orderKey`, so it should not reorder packets

Reply:
I think there is a race condition between deregistration of listeners on terminal job state (which is synchronous operation) and job state change event processing (when it is sent as async event not a sync operation). I could trigger this problem by injecting delay in `EventProcessor.process`. I lost terminal state change event but probably could loose more if the delay was large enough of the status change fast enough.

Reply:
in `ClientListenerServiceImpl.handleEventMessage`  job status events go into “Execute on a random worker” branch because partitionId=-1

Reply:
is get out-of-order messages in client, for example in `testListener_suspend_resume_restart_cancelJob` by adding `sleepMillis(100)` to `JobStatusLogger.jobStatusChanged` (at the beginning!)

Reply:
For the record: I checked if `DeregistrationOperation` can be reordered with events. It is executed first by operation thread before enqueueing deregistration in striped executor. However, it seem fine, because events are processed and enqueued in striped executor by IO thread, so race condition in this case is not possible.

Reply:
fixed with patch from @vbekiaris

Reply:
and added checking in tests if the invocations are on the same thread. It is not perfect, but enough to catch this error.

Reply:
@k-jamroz did we decide to make sure that the events are run on the same thread for the same job? if so, what about the client-side problem you wrote above?

Reply:
running on the same thread is the only way to ensure order apart from using sync events (which I expect would have terrible performance). Earlier @burakgok's changes ensure this on member side, patch from @vbekiaris (a3e91d1d30e55ecb66bb2607f986e5451a9f979a) fixed the client side ordering. So all problems with order seem to be fixed and not require sync events.

Reply:
Thanks, I have missed that commit 👍